### PR TITLE
ticlient: use arrays to store mutations

### PR DESCRIPTION
**Pull request by sticnarf**
## What problem does this PR solve? 

`map` has a poor memory efficiency. So it is a problem under large transactions that the `mutation` map in `twoPhaseCommitter` occupies too much memory.

### What is changed and how it works?

This PR uses an array to store the mutations, which is much more memory efficient. For correctness, this PR does not do too many reforms but a reimplementation of the same logic before.

This PR introduces some code repetition. I don't have a good idea for that, however.

### Check List 

Tests

 - Unit test
 - Integration test
 - Manual test

With this PR, the memory used for committing a 2.9G transaction decreases from 22G to 16G.

Side effects

 - Increased code complexity

Release note

 - Reduce memory usage of large transactions.

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM
Good job! @sticnarf

File: store/tikv/2pc.go [hunk 1]
```diff
@@ -480,29 +515,22 @@ func (c *twoPhaseCommitter) doActionOnBatches(bo *Backoffer, action twoPhaseComm
...
-	if mutation := c.mutations[string(key)]; mutation != nil {
-		size += len(mutation.Value)
-	}
-	return size
+func (c *twoPhaseCommitter) keyValueSize(key, value []byte) int {
+	return len(key) + len(value)
 }
 
-func (c *twoPhaseCommitter) keySize(key []byte) int {
+func (c *twoPhaseCommitter) keySize(key, value []byte) int {
```
Comment (jackysp) [inline store/tikv/2pc.go]: Why change it?

Comment (sticnarf) [reply store/tikv/2pc.go]: The grouping function now can handle both keys or mutations. So the size function takes both key and values. This change is to conform that signature change.

**Comment by jackysp** [comment]
Please resolve the conflicts, @sticnarf .

**Comment by sticnarf** [comment]
I‘ve resolved the conflicts. PTAL again.

**Pull request merged**